pkg/server/api/trustdomain.go

```diff
@@ -0,0 +1,124 @@
+package api
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire-api-sdk/proto/spire/api/types"
+	"github.com/spiffe/spire/pkg/common/protoutil"
+	"github.com/spiffe/spire/pkg/server/datastore"
+)
+
+// ProtoToFederationRelationship convert and validate proto to datastore federated relationship
+func ProtoToFederationRelationship(f *types.FederationRelationship) (*datastore.FederationRelationship, error) {
+	return ProtoToFederationRelationshipWithMask(f, nil)
+}
+
+// ProtoToFederationRelationshipWithMask convert and validate proto to datastore federated relationship, and apply mask
+func ProtoToFederationRelationshipWithMask(f *types.FederationRelationship, mask *types.FederationRelationshipMask) (*datastore.FederationRelationship, error) {
+	if f == nil {
+		return nil, errors.New("missing federation relationship")
+	}
+
+	if mask == nil {
+		mask = protoutil.AllTrueFederationRelationshipMask
+	}
+
+	trustDomain, err := spiffeid.TrustDomainFromString(f.TrustDomain)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse trust domain: %w", err)
+	}
+
+	var bundleEndpointURL *url.URL
+	if mask.BundleEndpointUrl {
+		bundleEndpointURL, err = url.Parse(f.BundleEndpointUrl)
```

Comment:
we should also ensure that the host is set... `url.Parse("https://")` will succeed